File: stations/hook-observatory/handlers/session_cost.py
"""
Session cost tracker — fire-and-forget response counter.

Appends one JSONL line per Stop event to track response counts per session.
All aggregation is deferred to the /session-cost command.

Latency: <1ms (atomic single-line append, no computation).
"""

from __future__ import annotations

import json
import os
from datetime import UTC, datetime

from .base import ALLOW, HOME, HookResult

DATA_DIR = os.path.join(HOME, ".claude", "data", "session-cost")
SESSIONS_FILE = os.path.join(DATA_DIR, "sessions.jsonl")

# Per-process counter: session_id → response_index
_counters: dict[str, int] = {}


def _parse_session_id(raw_input: str) -> str:
    """Extract session_id from raw hook payload. Returns empty string on failure."""
    try:
        parsed = json.loads(raw_input)
        # Claude Code Stop event embeds session_id at top level
        sid = parsed.get("session_id", "")
        if not sid:
            # Some payloads nest it under tool_input
            sid = parsed.get("tool_input", {}).get("session_id", "")
        return str(sid) if sid else ""
    except (json.JSONDecodeError, AttributeError):
        return ""


def handle(event_type: str, tool_name: str, tool_input: dict, raw_input: str) -> HookResult:
    """Track response count for each session. Always returns ALLOW."""
    if not raw_input.strip():
        return ALLOW

    try:
        session_id = _parse_session_id(raw_input) or "unknown"

        # Increment per-session counter (resets on process restart)
        _counters[session_id] = _counters.get(session_id, 0) + 1
        response_index = _counters[session_id]

        ts = datetime.now(tz=UTC).strftime("%Y-%m-%dT%H:%M:%SZ")
        line = json.dumps(
            {
                "session_id": session_id,
                "ts": ts,
                "tool_name": tool_name,
                "response_index": response_index,
            },
            ensure_ascii=False,
            separators=(",", ":"),
        )

        os.makedirs(DATA_DIR, exist_ok=True)
        with open(SESSIONS_FILE, "a") as f:
            f.write(line + "\n")
    except Exception:  # noqa: S110
        pass  # fire-and-forget: never block the hook pipeline

    return ALLOW

File: internal/handlers/session_cost.go
package handlers

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/joneshong/hook-dispatcher/internal/core"
)

func init() {
	core.Register("Stop", core.Entry{
		Handler:    sessionCostHandle,
		ModuleName: "session_cost",
	})
}

// sessionCostHandle is the Go port of handlers/session_cost.py.
//
// The Python version uses a process-scoped counter dict that starts empty on
// each dispatcher invocation, so response_index is effectively always 1 in
// practice (one dispatch per process). We replicate that behavior exactly.
func sessionCostHandle(_, toolName string, _ map[string]any, rawInput string) core.HookResult {
	if strings.TrimSpace(rawInput) == "" {
		return core.Allow()
	}

	sessionID := parseSessionID(rawInput)
	if sessionID == "" {
		sessionID = "unknown"
	}

	dataDir := core.Cfg().GetPath("data_dir")
	if dataDir == "" {
		home, _ := os.UserHomeDir()
		dataDir = filepath.Join(home, ".claude", "data")
	}
	costDir := filepath.Join(dataDir, "session-cost")
	if err := os.MkdirAll(costDir, 0o755); err != nil {
		return core.Allow()
	}
	sessionsFile := filepath.Join(costDir, "sessions.jsonl")

	ts := time.Now().UTC().Format("2006-01-02T15:04:05Z")
	entry := map[string]any{
		"session_id":     sessionID,
		"ts":             ts,
		"tool_name":      toolName,
		"response_index": 1, // matches Python per-process counter reset
	}
	line, err := json.Marshal(entry)
	if err != nil {
		return core.Allow()
	}

	f, err := os.OpenFile(sessionsFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
	if err != nil {
		return core.Allow()
	}
	defer f.Close()
	f.Write(line)
	f.Write([]byte("\n"))

	return core.Allow()
}

func parseSessionID(raw string) string {
	var parsed map[string]any
	if err := json.Unmarshal([]byte(raw), &parsed); err != nil {
		return ""
	}
	if v, ok := parsed["session_id"].(string); ok && v != "" {
		return v
	}
	if ti, ok := parsed["tool_input"].(map[string]any); ok {
		if v, ok := ti["session_id"].(string); ok && v != "" {
			return v
		}
	}
	return ""
}
